File: frontend/js/sim/kernel.js
// Fixed-step RK4 integrator (browser kernel). Integrates the same `derivatives`
// the Python reference uses; replaces PathSim for the client-side build. The
// system is tiny (6-8 ODEs) so plain JS runs it millions of steps/s — no WASM
// needed. Sub-steps internally for deterministic, stable integration.

export class Integrator {
  // Each model may declare its own micro-step (stiffer scenarios use a finer one)
  // and an optional state clamp (e.g. concentration >= 0). Defaults keep the
  // tank scenarios exactly as before.
  constructor(model) { this.model = model; this.dtMicro = model.dtMicro || 0.02; this.reset(); }

  reset(state) {
    this.x = (state || this.model.initialState()).slice();
    this.t = 0;
  }

  step(dt, act, env) {
    const m = this.model, f = (x) => m.derivatives(x, act, env);
    const nsub = Math.max(1, Math.ceil(dt / this.dtMicro - 1e-9)), h = dt / nsub;
    for (let s = 0; s < nsub; s++) {
      const x = this.x;
      const k1 = f(x);
      const k2 = f(x.map((v, i) => v + 0.5 * h * k1[i]));
      const k3 = f(x.map((v, i) => v + 0.5 * h * k2[i]));
      const k4 = f(x.map((v, i) => v + h * k3[i]));
      this.x = x.map((v, i) => v + (h / 6) * (k1[i] + 2 * k2[i] + 2 * k3[i] + k4[i]));
      if (m.clampState) this.x = m.clampState(this.x);
      this.t += h;
    }
    return m.buildState(this.x, act, env, this.t);
  }

  getState(act, env, t) { return this.model.buildState(this.x, act, env, t); }
}

File: aiogym/models/kernel.py
"""Fixed-step RK4 integrator — a line-for-line port of frontend/js/sim/kernel.js.

Each model declares its own micro-step (`dt_micro`); a control step `dt` is
sub-stepped into ceil(dt/dt_micro) RK4 micro-steps, and an optional `clamp_state`
is applied after every micro-step (e.g. concentration >= 0).
"""
from __future__ import annotations
import math


class Integrator:
    def __init__(self, model):
        self.model = model
        self.dt_micro = getattr(model, "dt_micro", 0.02) or 0.02
        self.reset()

    def reset(self, state=None):
        self.x = list(state) if state is not None else list(self.model.initial_state())
        self.t = 0.0

    def step(self, dt, act, env):
        m = self.model
        u = m.action_vector(act)
        f = lambda x: m.dynamics(x, u, env)
        nsub = max(1, math.ceil(dt / self.dt_micro - 1e-9))
        h = dt / nsub
        for _ in range(nsub):
            x = self.x
            k1 = f(x)
            k2 = f([v + 0.5 * h * k1[i] for i, v in enumerate(x)])
            k3 = f([v + 0.5 * h * k2[i] for i, v in enumerate(x)])
            k4 = f([v + h * k3[i] for i, v in enumerate(x)])
            self.x = [v + (h / 6.0) * (k1[i] + 2 * k2[i] + 2 * k3[i] + k4[i]) for i, v in enumerate(x)]
            if hasattr(m, "clamp_state"):
                self.x = m.clamp_state(self.x)
            self.t += h
        return self.x
